### DISCUSSION TITLE
useMutation to pause/resume mutations manually

### DISCUSSION THREAD

**Discussion by franferns** [TanStack/query#3580] [Q&A]
Hi,

using useMutation how do we solve a scenario where on calling of useMutation.mutate we can pause the mutation and resume when required. Currently pausing of mutation happens only when the device goes offline . 

I have a requirement for my react-native app saving data to the AsyncStorage that I have to persist mutation's offline even if the 
device is online and this persisted mutation will upload the data to the server calling the mutationFunction only on user interaction like click of an upload button for that mutation.

How can I achieve this ?

Thanks & Regards,
Francis P.

**Comment by TkDodo** [comment]
I don't quite understand the requirement - can you elaborate a bit? Mutations generally only trigger imperatively, e.g. on a button click...

**Reply by franferns** [reply]
The requirement is for a react-native app. The flow is as follows : 
1. The user answers a form with fields like firstname etc. these answers need to be persisted offline so the user opens the app next time the answered values are present .
2. The users should be able to upload the manually with click of Upload button . 

Probably my understanding was a little less of the library . But a solution would be to 

1. On submit handler queryClient.setQueryData(["questionnaire","1"],{id:"1",firstname:"user 1"})
2. Then use the queryClient.getQueryData hook to display/edit this saved data 
3. Use the useMutation hook when clicked on Upload button.
4. queryClient.removeQueries(["questionnaire","1"], { exact: true })

**Reply by TkDodo** [reply]
> The user answers a form with fields like firstname etc. these answers need to be persisted offline so the user opens the app next time the answered values are present .

The form state is _client state_. So whatever you use to manage that state should be responsible for persisting it.

Only after the user clicks the `upload` button, that state is passed to the mutation. Persisting mutations is only necessary if a mutation that already started  got aborted / failed due to network issues and the user quits the app _and_ you expect the mutation to retry once the user returns to the app.
